vumi_http_retry/limiter.py

```diff
@@ -0,0 +1,55 @@
+from twisted.internet.defer import (
+    Deferred, inlineCallbacks, DeferredLock, maybeDeferred)
+
+
+class TaskLimiter(object):
+    """
+    Keeps track of a limited list of pending tasks, allowing one to request for
+    a task to be added once there is a vacancy in the list.
+    """
+    def __init__(self, limit, errback=None):
+        self.lock = DeferredLock()
+        self.limit = limit
+        self.observers = []
+        self.tasks = set()
+
+        if errback is None:
+            errback = lambda f: f
+
+        self.errback = errback
+
+    @inlineCallbacks
+    def add(self, fn, *a, **kw):
+        observer = Deferred()
+        self.observers.append(observer)
+        observer.addCallback(self._run, fn, *a, **kw)
+        yield self._refresh()
+        yield observer
+
+    def _run(self, _, fn, *a, **kw):
+        d = maybeDeferred(fn, *a, **kw)
+        self.tasks.add(d)
+        d.addCallback(self._callback, d)
+        d.addErrback(self._errback, d)
```

Comment:
I think we could make things a bit clearer by combining `add` and `_run` like so:
```python
def add(self, fn, *a, **kw):
    d = Deferred()
    self.observers.append(d)
    self.tasks.add(d)
    d.addCallback(maybeDeferred, fn, *a, **kw)
    d.addCallbacks(self._callback, self._errback)
    return d
```